refactor!: GoogleOAuth

phot0n opened the pull request:
Changes:
- moved (common) google oauth stuff to a single class
- made a singular google callback as opposed to multiple whitelisting [breaking change] - except for google calendar
- added a generic oauth interface for emails (only enabled it for gmail - visible if the user selects service as `GMail`)
- added unique constraint on `email_id` field in email account doctype
- made `X-Original-From` header in email body consistent and non-changable even if the user chooses to change sender name/address
- added new field `Used Oauth` in `User Email` doctype to track awaiting password emails (except the ones who've used oauth)
- fixed field label inconsistencies in email account doctype
- fetched attachment_limit in email account from max_file_size config

#

Working:

<details>
  <summary>SMTP</summary>

https://user-images.githubusercontent.com/32034600/170973772-0d3f34b1-537f-4f40-a2a9-bed6a6a8d6c3.mov

</details>

<details>
  <summary>IMAP</summary>

https://user-images.githubusercontent.com/32034600/170973892-eeed96b5-565d-4fad-b2c1-6ed68038a2a7.mov

</details>

<details>
  <summary>POP3</summary>

https://user-images.githubusercontent.com/32034600/170974027-95d067c7-d686-4ebf-a749-e0fe7a4b0456.mov

</details>

#

TODO:
- [x] docs
- [x] check if `User Email` doctype needs to consider this (?)


will hopefully close: https://github.com/frappe/frappe/issues/14815

docs: 
- https://docs.erpnext.com/docs/v14/user/manual/en/erpnext_integration/google_settings/edit?wiki_page_patch=dd2d938a4e

phot0n committed:
refactor: GoogleOAuth

* refactor: single callback method for google oauth

phot0n committed:
feat: google oauth for google emails

* used unique constraint on email_id in Email Account Doctype

phot0n commented on frappe/core/doctype/communication/mixins.py:
```diff
@@ -94,7 +94,7 @@ def get_mail_cc_with_displayname(self, is_inbound_mail_communcation=False, inclu
 			is_inbound_mail_communcation=is_inbound_mail_communcation, include_sender=include_sender
 		)
-		return [self.get_email_with_displayname(email) for email in cc_list]
+		return [self.get_email_with_displayname(email) for email in cc_list if email]
```
was getting `None` email - hence added a check

phot0n commented on frappe/email/doctype/email_account/email_account.py:
```diff
@@ -496,6 +512,7 @@ def get_inbound_mails(self) -> List[InboundMail]:
 		def process_mail(messages, append_to=None):
 			for index, message in enumerate(messages.get("latest_messages", [])):
 				uid = messages["uid_list"][index] if messages.get("uid_list") else None
+				uid = uid.decode() if isinstance(uid, bytes) else uid
```
The uid was being sent as bytes 🤷‍♂️

![Screenshot 2022-05-27 at 12 26 30 AM](https://user-images.githubusercontent.com/32034600/170643987-99f5c6af-2634-42dd-8e99-97fc542b822b.png)

phot0n commented on frappe/email/utils.py:
```diff
@@ -15,3 +19,54 @@ def get_port(doc):
+			# poplib doesn't have AUTH command implementation
+			res = connection_obj._shortcmd(
+				"AUTH {0} {1}".format(mechanism, base64.b64encode(bytes(auth_string, "utf-8")).decode("utf-8"))
+			)
```
poplib doesn't have `AUTH` implementation - hence made a custom one 😛

phot0n commented on frappe/integrations/doctype/google_calendar/google_calendar.py:
```diff
@@ -14,7 +14,7 @@
 
 import frappe
```
Haven't touched google calender things (only updated the place from where oauth url was being fetched) as a refactor was going on for this.
ref: https://github.com/frappe/frappe/pull/15665

phot0n commented on frappe/core/doctype/communication/mixins.py:
```diff
@@ -122,7 +122,7 @@ def mail_bcc(self, is_inbound_mail_communcation=False):
 	def get_mail_bcc_with_displayname(self, is_inbound_mail_communcation=False):
 		bcc_list = self.mail_bcc(is_inbound_mail_communcation=is_inbound_mail_communcation)
-		return [self.get_email_with_displayname(email) for email in bcc_list]
+		return [self.get_email_with_displayname(email) for email in bcc_list if email]
```
was getting `None` email - hence added a check

phot0n commented on frappe/core/doctype/user/user.py:
```diff
@@ -751,19 +751,11 @@ def has_email_account(email):
 		fields=["email_account", "email_id"],
 		filters={"awaiting_password": 1, "parent": user},
 	)
-	if waiting:
-		return waiting
-	else:
-		user_email_table = DocType("User Email")
-		frappe.qb.update(user_email_table).set(user_email_table.user_email_table, 0).where(
-			user_email_table.parent == user
-		).run()
-		return False
```
this didn't make sense to me (?)
it's trying to update `tabUser Email.tabUser Email` - i think it was meant to be `awaiting_password` but then also it doesn't make sense as if we don't get any user emails with awaiting_password checked then it tries to set an already non-awaiting password user email to non-awaiting 🤷

phot0n commented on frappe/email/doctype/email_account/email_account.json:
```diff
@@ -79,7 +83,8 @@
    "options": "Email",
-   "reqd": 1
+   "reqd": 1,
+   "unique": 1
```
added unique constraint on `email_id` ..we are doing the same in `validate` hook

https://github.com/frappe/frappe/blob/d193ea22d17ca40d57432040a8afad72287d9e23/frappe/email/doctype/email_account/email_account.py#L84-L92

phot0n committed:
fix: allow adding new emails with google oauth checked

kinda works in a similar way as awaiting_password for new emails

surajshetty3416 commented on frappe/email/doctype/email_account/email_account.json:
```diff
@@ -18,6 +18,10 @@
   "awaiting_password",
   "ascii_encode_password",
   "column_break_10",
+  "use_google_oauth",
+  "authorize_google_api_access",
+  "google_refresh_token",
+  "google_access_token",
```
Can't these be general-purpose? What if we want to enable OAuth for another email service... 4 new fields?
```suggestion
  "use_oauth",
  "authorize_api_access",
  "refresh_token",
  "access_token",
```

phot0n replied:
hmm it can be done but then we need to have the authentication url as well which redirects us to the service page and the user allows/permits us to use their info.

kept it simple (only google) for the time being due to that (and the date for google stopping less secure apps is today so 🤷‍♂️ )

phot0n replied:
we could use the generic naming but only enable it for gmail for the time being as well (?)

phot0n committed:
fix: set X-Original-From header before replacing sender email/name

This will help in knowing the original email id and sender of email
and also bring consistency.

phot0n committed:
feat: generic OAuth for email

phot0n committed:
chore: fix linter

phot0n committed:
minor: throw exception if refresh_token is not present

phot0n committed:
chore: fix sider

phot0n committed:
minor: move oauth access functions from email_account

phot0n committed:
minor: fetch oauth fields from site config

phot0n committed:
minor: simplify authorize_google_access

phot0n committed:
chore: added docstrings

phot0n committed:
minor: better oauth flow

phot0n committed:
fix: dont update modified timestamp on generating and refreshing tokens

phot0n committed:
fix: open authorization url on the current browsing context

phot0n committed:
fix: use safe decode for uid

phot0n commented:
kinda don't like the UX - will update soon :P

phot0n committed:
fix: consider oauth usage as well for asking/updating user email password

gavindsouza commented on frappe/email/oauth.py:
```diff
@@ -0,0 +1,163 @@
+		)
+
+	res = oauth_obj.authorize(code, get_request_site_address(True))
+	frappe.db.set_value(
+		doctype, email_account, "refresh_token", res.get("refresh_token"), update_modified=False
+	)
+	frappe.db.set_value(
+		doctype, email_account, "access_token", res.get("access_token"), update_modified=False
+	)
```
This can be one query - pass it as a dict

gavindsouza commented on frappe/email/oauth.py:
```diff
@@ -0,0 +1,163 @@
+
+	if service == "GMail":
+		return authorize_google_access(email_account, doctype)
+
```
`raise NotImplementedError` for other services? 😆

gavindsouza commented on frappe/email/oauth.py:
```diff
@@ -0,0 +1,163 @@
+	Returns authorization url for redirection"""
+
+	if not service:
+		frappe.throw(frappe._("No Service is selected. Please select one and try again!"))
```
Why not just remove the default None?

gavindsouza commented on frappe/email/oauth.py:
```diff
@@ -0,0 +1,163 @@
+
+	@property
+	def _auth_string(self) -> str:
+		return "user=%s\1auth=Bearer %s\1\1" % (self.email, self._access_token)
```
f-strings more readable? 😅

gavindsouza commented on frappe/email/oauth.py:
```diff
@@ -0,0 +1,163 @@
+				res = self._connect_pop()
+
+				if not res.startswith(b"+OK"):
+					raise
```
raising bare exception? Maybe we should log more info here.

phot0n replied:
yep that's intentional ..it will get caught in the try except block for retrying/regenerating access token and if it again fails then we raise an OAuthenticationError

https://github.com/frappe/frappe/pull/16935/files#diff-0c840370456bfd1d0639a045bff918865f66d97116d7b05a7f9b06b1f1665232R76-R81

gavindsouza commented on frappe/integrations/doctype/google_drive/google_drive.py:
```diff
@@ -34,118 +29,56 @@ def validate(self):
-	authorize_access()
+	frappe.db.set_value("Google Drive", None, "authorization_code", oauth_code)
+	r = oauth_obj.authorize(oauth_code, get_request_site_address(True))
+	frappe.db.set_value("Google Drive", "Google Drive", "refresh_token", r.get("refresh_token"))
```
Since you're allowing GET, you may wanna add a commit down here. Also, single UPDATE query possible here

phot0n replied:
hmm we don't need GET here i suppose ..had moved it to a single place :P
https://github.com/frappe/frappe/pull/16935/files#diff-110d2f31c7a2be9ad9ee6b250d273ee033248ad49532149f1c473e21929124f3R167-R183

phot0n committed:
chore: remove GET method whitelisting from google integrations

phot0n committed:
chore: remove additional/unnecessary set_value calls

harshit-30 commented:
Any reason we haven't merged this yet? What's the ETA for this? 

@gavindsouza @surajshetty3416 @phot0n

phot0n commented:
> Any reason we haven't merged this yet? What's the ETA for this?

@harshit-30 This is my bad. Was on "kind of a break" the previous week - hence wasn't able to completely finish the pr (as you might've seen that there are some todo's pending).
Will Complete it within this week. Sorry for the delay.

phot0n committed:
fix: don't show authorize api access button if the form is unsaved

phot0n committed:
fix: remove commit

* chore: raise not implemented error for services other than gmail

* chore: use fstring for _auth_string property

phot0n committed:
chore: move getting site address to GoogleOAuth

phot0n committed:
fix: better/reduced exception handling for email oauth

Since the places where connection methods are called already
have a lot of exception handling, we can just raise and let them
handle all the probable cases.

harshit-30 commented:
@gavindsouza @surajshetty3416 can you please review it at your earliest?

surajshetty3416 commented:
> @gavindsouza @surajshetty3416 can you please review it at your earliest?

@harshit-30 We did review this over call... we'll get this merged once the requested changes are done.

surajshetty3416 commented:
@phot0n Any update on this? Let's get this merged ASAP.

phot0n committed:
minor: encrypt email oauth refresh and access token

phot0n committed:
minor(ux): select field for alternating between basic and oauth authentication

phot0n committed:
fix(ux): consistent field names for ssl and server

* keep server and port together

phot0n committed:
fix: remove fetch-from from user email for used_oauth

phot0n commented:
This is how it looks now (with a select field for selecting b/w basic and oauth auth - only shown for gmail service for now)

https://user-images.githubusercontent.com/32034600/178326015-3ec3f781-5d17-40ad-b711-08bbeee3fd4e.mov

harshit-30 commented:
@phot0n the updated video looks promising. Can you please notify the user that the sync has happened successfully? Currently there is no notification or toaster verifying that.

phot0n commented:
Updated the form header if refresh token is found.

https://user-images.githubusercontent.com/32034600/178460170-9db7cae1-01f0-427c-b922-fd1b3cd5ebf5.mov

phot0n committed:
fix(ux): update form header upon successful authorization

* minor: simplified validations for email account

phot0n committed:
minor: fetch attachment size from get_max_file_size api

phot0n committed:
chore: fix sider

surajshetty3416 commented on frappe/email/doctype/email_account/email_account.js:
```diff
@@ -118,30 +148,63 @@ frappe.ui.form.on("Email Account", {
+	show_oauth_authorization_message(frm) {
+		if (frm.doc.auth_method === "Oauth") {
+			let msg = {
+				message: !frm.doc.refresh_token ? "Oauth Enabled but not Authorized. Please use Authorize API Access Button to do the same." : "Oauth Authorized. Re-Authorization can be done using Authorize API Access Button.",
```
```suggestion
				message: __("OAuth has been enabled but not authorised. Please use "Authorise API Access" button to do the same."),
```
Show message only if not yet authorised?

phot0n replied:
had done that but as Harshit said that there should be some confirmation that the authorization happened - hence added a headline

Should i remove that?

surajshetty3416 replied:
Just show alert instead of permanent headline?

 `frappe.show_alert("Authorised successfully!")`

phot0n replied:
had used that initially, didn't really work - since we're not back in frappeland when the callback happens, it doesn't show anything 🤔

surajshetty3416 replied:
https://github.com/frappe/frappe/pull/16935/files#diff-0c840370456bfd1d0639a045bff918865f66d97116d7b05a7f9b06b1f1665232R162 
then `msgprint` from here?

surajshetty3416 replied:
resolved with https://github.com/frappe/frappe/pull/16935/commits/4a2e395adba5f1521ea3e416ca7e489d3029e358

surajshetty3416 reviewed (commented):
<img width="618" alt="Screenshot 2022-07-12 at 5 01 00 PM" src="https://user-images.githubusercontent.com/13928957/178480451-4de046af-58bc-4779-80a3-66dd5156b737.png">
Add link to navigate to "Google Settings" page?
The message can be 

**Please enable [Google Settings]() to authenticate via OAuth.**

phot0n committed:
fix: pull from accounts for oauth whose refresh_token is not null

* chore: rename Oauth to OAuth

phot0n committed:
chore: add link for google settings when throwing error

phot0n committed:
fix(ux): better ux for successful oauthorization

phot0n committed:
chore: use f-strings

phot0n committed:
chore: remove unnecessary failure_query_param for email oauth

phot0n committed:
chore: translate authorization message and add nosemgrep

surajshetty3416 committed:
Merge branch 'develop' into google_oauth_refactor
